### DISCUSSION TITLE
How do I exit interactive mode gracefully?

### DISCUSSION THREAD

**Discussion by xandfury** [Gallopsled/pwntools#2138] [Q&A] [answered]
I spawn a bash shell with the following args: ` env=None, setuid=True, stdin=PTY, raw=False`. A user exits the shell with `exit` and the bash process is exited. However, I am still in interactive mode. How does one exit interactive mode when the process has exited?

**Comment by peace-maker** [comment]
You can close the interactive shell by sending 0 bytes. You do that by pressing `ctrl+d` while your input line is empty to send an eof.

**Reply by xandfury** [reply]
Yep. `ctrl+d` and even `cntl+c` works. Was wondering if there was a more graceful way of exiting when `exit` is issued.

**Reply by peace-maker** [reply, marked as answer]
There's an open PR https://github.com/Gallopsled/pwntools/pull/2108 that closes the local input side when the remote connection dies, but it isn't done yet.
